torch/nn/parameter.py

```diff
@@ -44,3 +44,93 @@ def __reduce_ex__(self, proto):
         )
 
     __torch_function__ = _disabled_torch_function_impl
+
+
+class UninitializedParameter(Parameter):
+    r"""A parameter that can be lazily initialized.
+
+    Uninitialized Parameters holds empty tensors that can be moved
+    to different devices and change their type so when they are materialized,
+    they do it directly with the right configuration. Moreover, Uninitialized parameters
+    become full fledged :class:`torch.nn.Parameter` after they are initialized.
+    """
+    _allowed_methods = [
```

Comment:
This is a great idea. Should dunder getattr throw an error if the requested attribute isn't supported, too?

Reply:
Yes, we should definitely do that. However, I can't think of attributes that we need to forbid the access to right now 😅

Reply:
We can leave that for future work. Maybe a comment like "TODO: have dunder getattr throw an error when an unsupported attribute is requested" is all we need for now?